.azure-pipelines/collect_dump.py

```diff
@@ -0,0 +1,170 @@
+        members = tar.getmembers()
+        for member in members:
+            logger.info("Extracting {} {}".format(member.path, member.name))
+            if "log/syslog" in member.path:
+                try:
+                    logger.info("Extracting {} {} to {}".format(member.path, member.name, extract_path))
+                    tar.extract(member, path=extract_path)
+                except Exception as e:
+                    logger.info("Error extracting {} {}: {}".format(member.path, member.name, e))
+
+
+def extract_gz_file(gz_file_path):
+    extracted_file_path = os.path.splitext(gz_file_path)[0]
+    try:
+        with gzip.open(gz_file_path, 'rb') as f_in:
+            with open(extracted_file_path, 'wb') as f_out:
+                f_out.write(f_in.read())
+        os.remove(gz_file_path)
+    except Exception as e:
+        logger.info("Extract gz file {} failed: {}".format(gz_file_path, str(e)))
+        traceback.logger.info_exc()
+
+
+def extract_dump_file(testbed_name_with_idx, hostname):
+    try:
+        # extract dump file
+        dump_file = [file for file in os.listdir(LOGS_DIR)
+                     if file.startswith("sonic_dump") and file.endswith(".tar.gz")][0]
+        dump_file_path = os.path.join(TECHSUPPORT_SAVE_PATH, dump_file)
+        extract_dump_tar_gz(dump_file_path, TECHSUPPORT_SAVE_PATH)
+
+        # rename dump file
+        timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+        new_file_name = testbed_name_with_idx + "_" + hostname + "_" + timestamp
+        new_file_path = os.path.join(LOGS_DIR, new_file_name)
+        os.rename(os.path.join(_self_dir, dump_file_path), os.path.join(LOGS_DIR, new_file_name + ".tar.gz"))
+        os.rename(os.path.join(_self_dir, dump_file_path.split(".tar.gz")[0]), new_file_path)
+
+        # extract syslog gz files
+        syslog_dir = os.path.join(new_file_path, "log")
+        syslog_gz_files = [file for file in os.listdir(os.path.join(LOGS_DIR, syslog_dir))
+                           if file.startswith("syslog") and file.endswith(".gz")]
+        logger.info("Syslog files: {}".format(syslog_gz_files))
+        for syslog_gz in syslog_gz_files:
+            syslog_gz_files_path = os.path.join(syslog_dir, syslog_gz)
+            extract_gz_file(syslog_gz_files_path)
+
+    except Exception as e:
+        logger.info("Extract dump file failed: " + str(e))
+        traceback.logger.info_exc()
+
+
+def main(args):
+    logger.info("Initializing hosts")
+    sonichosts = init_testbed_sonichosts(
+        args.inventory, args.testbed_name, testbed_file=args.tbfile, options={"verbosity": args.verbosity}
+    )
+
+    if not sonichosts:
+        sys.exit(RC_INIT_FAILED)
+
+    if not os.path.exists(LOGS_DIR):
+        os.makedirs(LOGS_DIR)
+
+    for dut in sonichosts:
+        get_techsupport(dut, time_since=args.time_since)
+        extract_dump_file(testbed_name_with_idx=args.testbed_name_with_idx, hostname=dut.hostname)
```

Comment:
Extract and then upload the files one by one? This sounds very inefficient.

Reply:
fixed